From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 13 Jul 2010 12:05:00 -0400
Subject: [net] igb: drop support for UDP hashing w/ RSS
Message-id: <4C3C566C.8060300@redhat.com>
Patchwork-id: 26834
O-Subject: [RHEL 5.6 PATCH v2] igb: drop support for UDP hashing w/ RSS
Bugzilla: 613780
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Andy Gospodarek <gospo@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Bugzilla:
https://bugzilla.redhat.com/show_bug.cgi?id=613780

Description:
This change removes UDP from the supported protocols for RSS hashing. The
reason for removing this protocol is because IP fragmentation was causing a
network flow to be broken into two streams, one for fragmented, and one for
non-fragmented and this in turn was causing out-of-order issues.

v2: Properly indented, thanks Dean!

Upstream Status:
http://git.kernel.org/?p=linux/kernel/git/davem/net-next-2.6.git;a=commitdiff;h=4478a9cdf007a0418755a8a4016af8352fb1c1f3

Brew Build:
https://brewweb.devel.redhat.com/taskinfo?taskID=2590964

Test Status:
Tested by customer.

  Stefan

diff --git a/drivers/net/igb/igb_main.c b/drivers/net/igb/igb_main.c
index f60c042..f61d609 100644
--- a/drivers/net/igb/igb_main.c
+++ b/drivers/net/igb/igb_main.c
@@ -2250,15 +2250,16 @@ static void igb_configure_rx(struct igb_adapter *adapter)
 		for (j = 0; j < 10; j++)
 			array_wr32(E1000_RSSRK(0), j, random[j]);
 
-		mrqc |= (E1000_MRQC_RSS_FIELD_IPV4 |
-			 E1000_MRQC_RSS_FIELD_IPV4_TCP);
-		mrqc |= (E1000_MRQC_RSS_FIELD_IPV6 |
-			 E1000_MRQC_RSS_FIELD_IPV6_TCP);
-		mrqc |= (E1000_MRQC_RSS_FIELD_IPV4_UDP |
-			 E1000_MRQC_RSS_FIELD_IPV6_UDP);
-		mrqc |= (E1000_MRQC_RSS_FIELD_IPV6_UDP_EX |
-			 E1000_MRQC_RSS_FIELD_IPV6_TCP_EX);
-
+		/*
+		 * Generate RSS hash based on TCP port numbers and/or
+		 * IPv4/v6 src and dst addresses since UDP cannot be
+		 * hashed reliably due to IP fragmentation
+		 */
+		mrqc |= E1000_MRQC_RSS_FIELD_IPV4 |
+			E1000_MRQC_RSS_FIELD_IPV4_TCP |
+			E1000_MRQC_RSS_FIELD_IPV6 |
+			E1000_MRQC_RSS_FIELD_IPV6_TCP |
+			E1000_MRQC_RSS_FIELD_IPV6_TCP_EX;
 
 		wr32(E1000_MRQC, mrqc);
 
